src/fhirtypes/StructureDefinition.ts

```diff
@@ -200,6 +253,95 @@ export class StructureDefinition {
+        // Get the bracket elements and slice off the trailing ']'
+        const brackets = splitPathPart.slice(1).map(s => s.slice(0, -1));
+        pathParts.push({ base: fhirPathBase, brackets: brackets });
+      }
+    }
+    return pathParts;
+  }
+
+  /**
+   * Looks for a matching choice element and if found slices it, adds the slices to the structdef
+   * and then returns the newly created slice.
+   * @param {string} fhirPath - The path in FHIR to match with
+   * @param {ElementDefinition[]} elements - The set of elements to search through
+   * @returns {ElementDefinition} - The new slice element if found, else undefined
+   */
+  private sliceMatchingValueX(fhirPath: string, elements: ElementDefinition[]): ElementDefinition {
+    let matchingType: ElementDefinitionType;
+    const matchingXElement = elements.find(e => {
+      if (e.path.endsWith('[x]')) {
+        for (const t of e.type) {
+          if (`${e.path.slice(0, -3)}${capitalize(t.code)}` === fhirPath) {
+            matchingType = t;
+            return true;
+          }
+        }
+      }
+    });
+    if (matchingXElement) {
+      // If we find a matching [x] element, we need to slice it to create the child element
+      matchingXElement.sliceIt('type', '$this', false, 'open');
+      // Get the sliceName for the new element
+      const sliceName = fhirPath.slice(fhirPath.lastIndexOf('.') + 1);
+      const newSlice = matchingXElement.addSlice(sliceName, matchingType);
+      return newSlice;
+    }
+    return;
+  }
+
+  /**
+   * Looks for a slice within the set of elements that matches the fhirPath
+   * @param {PathPart} pathPart - The path to match sliceName against
+   * @param {ElementDefinition[]} elements - The set of elements to search through
+   * @returns {ElementDefinition} - The sliceElement if found, else undefined
+   */
+  private findMatchingSlice(pathPart: PathPart, elements: ElementDefinition[]): ElementDefinition {
+    return elements.find(e => e.sliceName === pathPart.brackets.join('/'));
+  }
+
+  /**
+   * Looks for a Reference type element within the set of elements that matches the fhirPath
+   * @param {PathPart} pathPart - The path to match the Reference type elements against
+   * @param {ElementDefinition[]} elements - The set of elements to search through
+   * @returns {ElementDefinition} - The Reference type element if found, else undefined
+   */
+  private findMatchingRef(pathPart: PathPart, elements: ElementDefinition[]): ElementDefinition {
+    const matchingRefElement = elements.find(e => {
+      // If we have foo[a][b][c], and c is the ref, we need to find an element with sliceName = a/b
+      if (
+        pathPart.brackets.length === 1 ||
+        e.sliceName === pathPart.brackets.slice(0, -1).join('/')
+      ) {
+        for (const t of e.type) {
+          return (
+            t.code === 'Reference' &&
+            t.targetProfile &&
+            t.targetProfile.find(tp => tp.endsWith(pathPart.brackets[0]))
```

Comment:
In theory, this could catch more than it should.  For example:
```
foo[Group]
```
should find a reference to a `Group` -- but would also match on a reference to a `RequestGroup` (which is actually a very different thing).

So... I think you need something more like:
```typescript
t.targetProfile.find(tp => tp === pathPart.brackets[0] || tp.endsWith(`/${pathPart.brackets[0]}`))
```

Reply:
I see you went with just the `endsWith` part.  Since this is a `canonical` I think that is OK, but in past versions of FHIR, it was possible for this just to be resource name, like `"Condition"`.

Also, note that a `canonical` _can_ have the version number appended to the end (delimited by `|`).  We're not supporting that here.  It's probably ok for now, but we may have to come back to it.  And we'll need to decide if we just ignore the version in that case or if we need to do something else special.

Reply:
My mistake, I can add the other check for just a resource name. And then I will add a comment for the version number situation.

Reply:
Actually, I'm afraid this might come back and bite us at exactly the wrong moment.  How about we fix it so it will work even when there is a version at the end -- and for now we ignore the version.  So... for example, a path with `[us-core-practitioner]` should match on a type with `"profile": "http://hl7.org/fhir/us/core/StructureDefinition/us-core-practitioner|3.0.0"`.

Reply:
Ok I will happily match whether the version number is there or not, but ignore the version number.

Reply:
Oooh... _happily_!  I'm glad to have made your morning!

Reply:
This should be addressed here: https://github.com/standardhealth/sushi/pull/1/commits/b9dc27051fe4e9a6fb3f28f031fb3dce8a6946a0.